core/src/commonJvmMain/kotlin/com/ustadmobile/core/torrent/ContainerTorrentDownloadJob.kt

```diff
@@ -0,0 +1,218 @@
+        val containerUid = contentJobItem.cjiContainerUid
+
+        // check if torrent file already exists
+        val torrentFile = File(torrentDir, "$containerUid.torrent")
+        return withContext(Dispatchers.Default) {
+
+            try {
+
+                if (!torrentFile.exists()) {
+                    val path = UMFileUtil.joinPaths(endpoint.url, "containers/$containerUid")
+                    val torrentFileStream = httpClient.get<InputStream>(path)
+                    torrentFile.writeBytes(torrentFileStream.readBytes())
+                }
+
+                val downloadFolderPath = jobItem.contentJob?.toUri?.let { URI(it).path }
+                        ?: containerDir.path
+
+                val containerFilesFolder = File(downloadFolderPath, containerUid.toString())
+                containerFilesFolder.mkdirs()
+
+                val metadataProvider = FileMetadataProvider(torrentFile.absolutePath)
+                val fileNameList = metadataProvider
+                        .torrentMetadata
+                        .files
+                        .map { File(it.relativePathAsString).name }
+
+                val existingFiles = db.containerEntryFileDao.findEntriesByMd5SumsSafeAsync(fileNameList,
+                        db.maxQueryParamListSize)
+
+                //link them to the new container path
+                existingFiles.forEach {
+                    val existingPath = it.cefPath ?: return@forEach
+                    val existingFile = File(existingPath)
+                    if (existingFile.name == MANIFEST_FILE_NAME) {
+                        return@forEach
+                    }
+                    val target = File(containerFilesFolder, existingFile.name)
+                    if (!target.exists()) {
+                        createSymLink(existingPath, target.path)
+                    }
+                }
+
+                withWifiLock(context) {
+
+                    val startTime = System.currentTimeMillis()
+                    ustadTorrentManager.addTorrent(containerUid, containerFilesFolder.path)
+
+                    val torrentDeferred = CompletableDeferred<Boolean>()
+                    val torrentListener = object : TorrentDownloadListener {
+                        override fun onComplete() {
+                            contentJobItem.cjiItemProgress = (contentJobItem.cjiItemTotal * 0.75).toLong()
+                            progress.onProgress(contentJobItem)
+                            println("downloaded torrent in ${System.currentTimeMillis() - startTime} ms")
+                            torrentDeferred.complete(true)
+                        }
+
+                        override fun onProgress(progressSize: Int) {
+                            contentJobItem.cjiItemProgress = (progressSize * 0.75).toLong()
+                            progress.onProgress(contentJobItem)
+                        }
+                    }
+                    GlobalScope.launch(Dispatchers.Default) {
```

Comment:
This is creating a leak! We should use launch without GlobalScope. It should be in the same scope, so that if the context is canceled, this should also be canceled. The isActive in this loop refers to the new GlobalScope context, so it will always be true. 

This code is actually not needed. Await will throw the cancellation exception if it is canceled.

Then (as per comment below) we need to make sure that when the cancellation exception is thrown, we remove the torrent. 

Perhaps UstadTorrentManager's removeTorrent function should take a parameter to indicate whether or not to delete the actual torrent file. Then the job can call removeTorrent and indicate that it should not delete the torrent file itself.